src/mbgl/vulkan/context.cpp

```diff
@@ -0,0 +1,537 @@
+    return std::make_unique<DrawableBuilder>(std::move(name));
+}
+
+gfx::UniformBufferPtr Context::createUniformBuffer(const void* data, std::size_t size, bool persistent) {
+    return std::make_shared<UniformBuffer>(createBuffer(data, size, VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT, persistent));
+}
+
+gfx::ShaderProgramBasePtr Context::getGenericShader(gfx::ShaderRegistry& shaders, const std::string& name) {
+    const auto shaderGroup = shaders.getShaderGroup(name);
+    auto shader = shaderGroup ? shaderGroup->getOrCreateShader(*this, {}) : gfx::ShaderProgramBasePtr{};
+    return std::static_pointer_cast<gfx::ShaderProgramBase>(std::move(shader));
+}
+
+TileLayerGroupPtr Context::createTileLayerGroup(int32_t layerIndex, std::size_t initialCapacity, std::string name) {
+    return std::make_shared<TileLayerGroup>(layerIndex, initialCapacity, std::move(name));
+}
+
+LayerGroupPtr Context::createLayerGroup(int32_t layerIndex, std::size_t initialCapacity, std::string name) {
+    return std::make_shared<LayerGroup>(layerIndex, initialCapacity, name);
+}
+
+bool Context::emplaceOrUpdateUniformBuffer(gfx::UniformBufferPtr& buffer,
+                                           const void* data,
+                                           std::size_t size,
+                                           bool persistent) {
+    if (buffer) {
+        buffer->update(data, size);
+        return false;
+    } else {
+        buffer = createUniformBuffer(data, size, persistent);
+        return true;
+    }
+}
+
+gfx::Texture2DPtr Context::createTexture2D() {
+    return std::make_shared<Texture2D>(*this);
+}
+
+RenderTargetPtr Context::createRenderTarget(const Size size, const gfx::TextureChannelDataType type) {
+    return std::make_shared<RenderTarget>(*this, size, type);
+}
+
+std::unique_ptr<gfx::OffscreenTexture> Context::createOffscreenTexture(Size size,
+                                                                       gfx::TextureChannelDataType type,
+                                                                       bool depth,
+                                                                       bool stencil) {
+    return std::make_unique<OffscreenTexture>(*this, size, type, depth, stencil);
+}
+
+std::unique_ptr<gfx::OffscreenTexture> Context::createOffscreenTexture(Size size, gfx::TextureChannelDataType type) {
+    return createOffscreenTexture(size, type, false, false);
+}
+
+std::unique_ptr<gfx::TextureResource> Context::createTextureResource(Size,
+                                                                     gfx::TexturePixelType,
+                                                                     gfx::TextureChannelDataType) {
+    assert(false);
+    return nullptr;
+}
+
+std::unique_ptr<gfx::RenderbufferResource> Context::createRenderbufferResource(gfx::RenderbufferPixelType, Size) {
+    return nullptr; // std::make_unique<RenderbufferResource>();
```

Comment:
todo, or extra comment?

Reply:
Could you clarify @adrian-cojocaru ?

Reply:
Initially it was a TODO. It ended up as a `nullptr` in all the tests revealing that the RenderbufferResource is never used.